Added Template feature so documents can be generated using the built in functions

Mrpye opened the pull request:
Not sure if this is something you want to add, but I have found it a very useful feature, so spent the time making it a feature.

- Added the ability to generate forms using Go's template engine.
- You can pass in a template with the functions and it populates a document.
- You can also specify the number of lines if your template has a range.
```
{{range $y := IntRange 1 .Lines}}
```
# example
```
http://localhost:8080/template?template="{{Name}} {{LastName}}"&lines=3

output:
Markus Moen Pagac
Matteo    Krajcik
Modesta   Hilpert
```

- Added 4 data templates email, fixed_format, markdown and html.

```
http://localhost:8080/template?template_type="fixed_format"&lines=3

output:
FIRSTNAME LASTNAME  Gender  Job            Hobby               CreditCard          Address
Markus    Moen      male    Developer      Lacrosse            39800889982276      369 North Cornerbury, Miami, North Dakota 24259
Matteo    Krajcik   female  Manager        Tour skating        569586614968023     4090 Lake Unionport, Pittsburgh, Iowa 73720
Modesta   Hilpert   female  Representative Volleyball          6495603064524543520 976 Forgeport, Aurora, Minnesota 92470
```

I have created unit tests for the template and functions.

To make the functions available to the template engine there is a function called **BuildFunctionMap**, that gets the methods from **globalFaker** and checks that the parameters are compatible with GO's template engine, and then adds them to a function map.

See what you think,

Thanks Andrew

brianvoe commented on README.md:
```diff
@@ -269,12 +269,13 @@ All functions also exist as methods on the Faker struct
 ### File
 
-Passing `nil` to `CSV`, `JSON` or `XML` it will auto generate data using a random set of generators.
+Passing `nil` to `CSV`, `JSON`, `Template` or `XML` it will auto generate data using a random set of generators.
```
swap template and xml as template is an outlier compared to the others

brianvoe commented on README.md:
```diff
@@ -269,12 +269,13 @@ All functions also exist as methods on the Faker struct
 CSV(co *CSVOptions) ([]byte, error)
 JSON(jo *JSONOptions) ([]byte, error)
 XML(xo *XMLOptions) ([]byte, error)
+Template(xo *TemplateOptions) ([]byte, error)
```
new line to separate please

Mrpye replied:
just checking is the new line after Template?

brianvoe commented on data/template.go:
```diff
@@ -0,0 +1,113 @@
+</html>`
+
+// replaceSingleQuote replaces single quote with multiline quote
+func GetTemplateData(template string) string {
```
Instead of one function and we pass in a string I think we should do each function separately with the preset "Template"

Ex: TemplateEmail, TemplateHtml, etc...

brianvoe commented on lookup.go:
```diff
@@ -139,6 +139,7 @@ func initLookup() {
 	addFileJSONLookup()
 	addFileXMLLookup()
 	addFileCSVLookup()
+	addFileTemplateLookup()
```
addTemplateLookup

brianvoe commented on template.go:
```diff
@@ -0,0 +1,270 @@
+
+}
+
+func addFileTemplateLookup() {
```
I would assume not all will be files so we can just name it addTemplateLookup

brianvoe commented on template.go:
```diff
@@ -0,0 +1,270 @@
+		return nil, err
+	}
+
+	//**************************************
```
Good comments but we dont need the star ones.

brianvoe commented on template.go:
```diff
@@ -0,0 +1,270 @@
+	}
+
+	//******************
+	//Return the result
```
comment probably not necessary.

brianvoe commented on template.go:
```diff
@@ -0,0 +1,270 @@
+		co.Template = data.GetTemplateData(co.TemplateType)
+	}
+
+	//Sort out the lines to pass to the template
```
space at beginning.

brianvoe commented on template.go:
```diff
@@ -0,0 +1,270 @@
+	}
+	td := templateData{Lines: co.Lines}
+
+	//check if we have a function map
```
space at beginning.

brianvoe commented on template_test.go:
```diff
@@ -0,0 +1,237 @@
+)
+
+func ExampleTemplate() {
+	//Make sure we get the same results every time
```
space at beginning.

brianvoe commented on template_test.go:
```diff
@@ -0,0 +1,237 @@
+}
+
+func ExampleFaker_Template() {
+	//Make sure we get the same results every time
```
space at beginning.

brianvoe commented:
Wow! I really like the direction of your thought process on this. I definitely think this idea is something we can get in.

Im going to try to do my best to describe my thoughts on adjustments I think you should make that would get this integrated the best.

1. Template - So the first one I think we should tackle is simplifying it down to only the primary functions that people will run and not having a function that passes a string to lookup a template. Instead lets just keep it simple and list out the 5 functions you want. Ex: Template(Love your first example!) TemplateEmail, TemplateMarkdown, TemplateHtml and TemplateFixedFormat(Might want to think maybe about a different name, as this might clash with a future development of static generators in the future. Or i can be convinced with more example usages.)
2. Public functions - We have to be super careful about making sure we only create a small necessary amount of public functions. The one that pops up is the BuildFunctionMap. If the function purpose is for the use of templates only then we just make them private. I dont normally make general helper functions that are generally only used for a gofakeit function public.
3. Misc functions - If they are general functions that the whole package can use I put them in the helpers file and make sure they have specific test for it. Minor thing but it helps organize this and removes potential similar functions from being created.

I added some comments in your files to start off. We can keep going back and fourth on adjustments and ideas and well see where we end on this.

Thanks again and nice job.

brianvoe commented on template.go:
```diff
@@ -0,0 +1,270 @@
+	"github.com/brianvoe/gofakeit/v6/data"
+)
+
+// CSVOptions defines values needed for csv generation
```
comment update

brianvoe reviewed (commented):
First quick round of comments.

Mrpye commented:
I am glad you like the idea, I thought it would be useful
This is my first contribution to another project.

I'll take a look and make changes I like the idea of separate functions for each template type.

Thanks for the feed back and ill get to work

brianvoe commented:
Sounds good. Good luck!

brianvoe commented:
And make sure tests are passing too. :)

Mrpye commented:
The FixedFormat name, how about 'Fixed Width' seems to be an official name for that format.

brianvoe commented:
That template is tricky, I dont know if i fully see the benefit of that one. My thought with that one is instead of fixed with in the main template function you would have an option of fixed width and header. And with that I feel like it should tackle what the fixedformat one was trying to accomplish. Not sure if that makes sense but we can always swing back around to that one and figure something out.

Mrpye commented:
I have removed the fixed width for now,  will revisit yes it perhaps should work similar to  the csv function.

Mrpye commented:
I am just adding better html template data. I have basic page and multiple content items that get randomly generated depending on how many sections you pass.
Sro we now have
TemplateDocument
TemplateEmail
TemplateMarkdown
TemplateHtml(sections int)
TemplateHtmlContent

Unfortunately I managed to injury myself,  so lost a few days in hospital but back on it now.

Mrpye commented:
Hi @brianvoe,

How did you feel about using the go:embed? I am only asking because it feels messy adding templates as code, having them as files means keeping the code clean and it's easier to manage and edit files.

I thought I would ask the question, but I am happy to go with what you feel is best.

Andrew

brianvoe commented:
Glad to see your doing well. Welcome back. 

Lets start by getting it cleaned up and perfected on a simple scale then it makes it easier when you want to add more.

Lets stay away from go:embed. Just makes it more difficult to use the package. String templates in a Go file are perfectly fine.

Let me know if you need anything. Thanks!

Mrpye commented:
Summary of updates

Split the Template function into individual functions

Template(template string, lines int) ([]byte, error) // Custom templates
TemplateDocument() (string, error) // Random Document from HTML,markdown,Email
TemplateHtml(sections int) (string, error) //Email Document
TemplateHtmlContent() (string, error) // HTML Content
TemplateMarkdown(sections int) (string, error) // markdown Document
TemplateMarkdownContent() (string, error) // markdown content
TemplateEmail() (string, error)  // Email text and HTML

Moved the Template functions in the readme to its own Template section 

Added a couple of different templates to the data 

Added Unit tests for each function and also a big test to make sure all the templates compile properly

Several other changes

brianvoe commented on .vscode/launch.json:
I dont add other peoples settings to my repos

brianvoe commented on README.md:
```diff
@@ -768,3 +769,18 @@ ErrorRuntime() error
+Template(template string, lines int) ([]byte, error)
+TemplateDocument() (string, error)
+TemplateHtml(sections int) (string, error)
+TemplateHtmlContent() (string, error)
```
Lets just stick with the main for now. So no Content ones

brianvoe commented on README.md:
```diff
@@ -768,3 +769,18 @@ ErrorRuntime() error
+TemplateHtml(sections int) (string, error)
+TemplateHtmlContent() (string, error)
+TemplateMarkdown(sections int) (string, error)
+TemplateMarkdownContent() (string, error)
```
Same as above

brianvoe commented:
Ok that was a good round of changes. Lets talk about the next set of updates. I like your ambition for having "full" template but the templates you have are really bloated. Think MUCH simpler. Try to remove all the extra stuff. The goal is to get as broad amount of normal usage as possible and leave it open for others to build upon if they have more demanding situations(hence why just template where they can add there own and generate the amount of lines is a nice one)

For email just have your basic title, message, signature and get really creative in those locations.

- title - randomly select (hi, hello, etc...) {{name}},
- content - randomly build sentences, questions, paragraphs. Mix them in together to make is feel various.
- signature - randomly select a goodbye type ending. whether or not to have thanks or your contact info in it.

Markdown looks a little better but instead of multiple "templates" think one template and randomly generating certain areas to fill out.

The html is a tricky one. I would say leave that out for now as I have a ticket for the one specifically and that will need a different level of heaving building to get that one right.

Hopefully that helps and look forward to your updates

brianvoe commented on .vscode/settings.json:
```diff
@@ -1,13 +0,0 @@
-{
```
Please dont delete my files

brianvoe commented on template.go:
```diff
@@ -0,0 +1,437 @@
+}
+
+// function to build the function map for the template engine from the global faker
+func createGlobalFakerFunctionMap() template.FuncMap {
```
templateFuncMap

brianvoe commented on template.go:
```diff
@@ -0,0 +1,437 @@
+	}
+
+	// function to generate a base64 encode string useful for images
+	funcMap["Base64Enc"] = base64EncString
```
remove for now

brianvoe commented on template.go:
```diff
@@ -0,0 +1,437 @@
+	funcMap["uc"] = strings.ToUpper
+
+	// function wrapper for SVG this is because template engine cant handle passing structs
+	funcMap["SVG"] = func(width int, height int) string {
```
remove for now. Cant just use globalFaker cause we need to use the rand that was used in the function request.

brianvoe commented:
Looks much better!!!

I think the next round is:

1. making sure things in the templates are general and not specific to any one thing in particular(ie:  golang). I know this is a golang package but I have usages for command line and servers that wont specifically pertain to golang developers. Plus just cause a golang developer is using it doesnt necessarily mean thats the output they want.
2. createGlobalFakerFunctionMap - rename to templateFuncMap
3. createGlobalFakerFunctionMap - I understand the purpose of it but want to see how much we can drop down to, to only include the essentials.
4. createGlobalFakerFunctionMap - remove base64Enc, lets not do images for template right now.
5. createGlobalFakerFunctionMap - the keys in the func map need to be the same case.
6. No need for var Template as you can just use the template string in the function that uses it.
7. Would love to see a small but significant example usage in the readme. Ive done this in the past for things like random, struct and fakable types. Just enough so people understand its potential.

You are super close. Keep up the good work!

brianvoe commented on README.md:
```diff
@@ -768,3 +769,15 @@ ErrorRuntime() error
+Template(template string, lines int) ([]byte, error)
+TemplateDocument(sections int) (string, error)
+TemplateMarkdown(sections int) (string, error)
+TemplateEmail(sections int) (string, error) 
```
TemplateEmailText - lets change it to TemplateEmailText for now in the event we do an html version in the future.

Mrpye commented:
## I understand the purpose of it but want to see how much we can drop down to, to only include the essentials.
In regards to reducing the functions in templateFuncMap i'll justify each function then you decide if it needs to go

- **Keep funcMap["IntRange"]** we need it we need to create loops range $y := IntRange 1 10
-    **Keep funcMap["CreateListResult"] and funcMap["ListResult"]** would keep this these two functions allow us to randomize a range of values and the
    Each loop removes a value from the list. This is useful so when randomizing content we don't repeat values until the list is exhausted the fresh range is generated

- would **keep** these Some basic string manipulation functions
**funcMap["Replace"]** // function to replace strings in strings
**funcMap["LCase"]** // function to make string lower case
**funcMap["Concat"]** // function to concatenate strings
**funcMap["UCase"]**  // function to make string upper case

- **funcMap["ListI"]** this to use the Weighted function 
- **funcMap["ListS"]** this to use the RandomString and Teams function 
- **funcMap["ListUInt"]** this to use the Dice and RandomUint function 
- **funcMap["ListInt"]** this to use the RandomInt function 
- 
## Possibility to prune

- funcMap["ListF32"]  function to enable passing slice of float32 to functions not needed any funcs I can see
- Not sure how useful these are in the context of the template
RandomMapKey
RandomMapKey
if you don't use the above in the context of the template then we can get rid of these
funcMap["MapSInt"]
funcMap["MapIntS"]

Not sure how useful this would be so can possible go
funcMap["Sql"]

brianvoe commented:
Lets prune the "Possible prune" ones.

For some of these other ones lets just use the same name as the function that uses it.
funcMap["ListI"] this to use the Weighted function
funcMap["ListS"] this to use the RandomString and Teams function
funcMap["ListUInt"] this to use the Dice and RandomUint function
funcMap["ListInt"] this to use the RandomInt function

For LCase and UCase we can just use Upper and Lower

I need to think about these one a little so we can revaluate later. funcMap["CreateListResult"] and funcMap["ListResult"]

Mrpye commented:
# No need for var Template as you can just use the template string in the function that uses it.

Can you clarify on this point?

The TemplateDocument() uses  getRandValue to get a random document var Template

Mrpye commented:
I see your point 

In that case, just have 3 main functions

Template
TemplateMarkdown
TemplateEmailText


>No need for var Template as you can just use the template string in the function that uses it.

- remove var Template
```
var Template = map[string][]string{
	"email":    {email_template1},
	"markdown": {markdown_body1},
	"document": {email_template1, markdown_body1},
}
```

- Am I keeping these in the data? and then reference them in my code
```
// Email Template
const email_template1 = `...` 

// Markdown Template
const markdown_body1 = `...`
```

or just move them my main code?

Mrpye commented:
if I keep them in data I would have to make them public

```
// Email Template
const Template_email_text = `...` 

// Markdown Template
const Template_markdown = `...`
```

brianvoe commented:
You can put them right outside the function in a private variable.

brianvoe commented:
Its beautiful!

But, I cant add them much height to the readme. We gotta try to figure out how to cut this in half. You could probably cut the template usage section and just add the golang link to the main description. And then your example usage try to figure out parts you can cut that basically are already describe in normal golang templates. Lets start there and see what it looks like. I think with those two things minimized it should still get your point across in a smaller overall height.

Side Note: Sorry i didnt notice this before but I think you should make the TemplateMarkdown and TemplateEmailText take in a struct of options that way you have way more flexibility to add more options in the future.

Nice job!

Mrpye commented:
Version 2 documentation

### Template

Generate custom documents using golang's template engine [https://pkg.go.dev/text/template](https://pkg.go.dev/text/template).

```go
Template(co *TemplateOptions) ([]byte, error)
TemplateMarkdown(co *TemplateOptions) (string, error)
TemplateEmailText(co *TemplateOptions) (string, error) 
```

#### Template Helper Functions

To help with formatting and using Gofakeit function there are some template helper functions available.

<details>
  <summary>1. String Functions</summary>

```go
- Replace(s string, old string, new string) string // Replace a old string with new string
- Concat(sep string,args ...string) string // concatenate strings together using a separator
- Upper(s string) string // make string upper case
- Lower(s string) string // make string lower case
```

</details>

<details>
  <summary>2. Slice Functions</summary>

```go
- Replace(s string, old string, new string) string // Replace a old string with new string
- Concat(sep string,args ...string) string // concatenate strings together using a separator
- Upper(s string) string // make string upper case
- Lower(s string) string // make string lower case
```

</details>

<details>
  <summary>3. Unavailable Gofakeit functions</summary>

The following Gofakeit function are not available to use in templates

```go
- RandomMapKey(mapI interface{}) interface{}
- ShuffleAnySlice(v interface{})
- ShuffleInts(a []int)
- ShuffleStrings(a []string)
- Struct(v interface{})
- Slice(v interface{})
- Svg(options *SVGOptions) string
```

</details>


#### Example Usage
```go
import "github.com/brianvoe/gofakeit/v6"

// Accessing the Lines variable from within the template.
template:=`{{range $y := IntRange 1 .Lines}}`

// Example of using a helper functions to build a List of strings and pass it to a gofakeit function, Nested function need to be wrapped in ().
template +=`{{RandomString (ListS "Contact Details" "Customers Details")}}{{$p:=Person}}` 

// Displaying values from a variable, and using Upper helper function to format Company name.
template +=`Name: {{$p.FirstName}} {{$p.LastName}}/nCompany: {{Upper ($p.Job.Company)}}/nContact: {{$p.Contact.Email}}{{end}}`

value, err := gofakeit.Template(template, 4)
if err != nil {
	fmt.Println(err)
}

fmt.Println(string(value))

//Output:
// Contact Details
// 
// Name: Philip Casper
// Company: POSSIBILITYU
// Contact: jerodhilll@tillman.name
```

brianvoe commented:
Perfect!!! Let me give everything one more once over and ill get this released

brianvoe commented:
Looks like you still have some code changes that need to be made.

Ill wait till your ready

brianvoe commented:
Good Morning,

I had an ah ha moment this morning. Im going to layout my thoughts and you let me know what you think.

// Templates
Template(template string, options TemplateOptions)
EmailText(options EmailOptions)
Markdown(options MarkdownOptions)
FixedWidth(options FixedWidthOptions)

We can get rid of the word Template in front of the function name. They are all still in the category of template but just dont need to say it. I dont know why i didnt realize it sooner, but most other functions dont need to preceed the function name with the category.

For the main template let them define the lines if they want to in there own template. I think the issue I have with lines being a param is I can think of a bunch of scenarios that I would want to output a template that dont require lines. That being said I feel like it should take in a second param for options and if they want to pass in lines as a dataset to be used in there template they can.

```go
type TemplateOptions struct {
    Data interface{}
    Funcs template.FuncMap
}
```

I want to get your fixed width back in and i think the original issue is we were the ones defining the template instead of just letting them pass in the functions they want to use.

```go
type FixedWidthOptions struct {
    Header []string
    Funcs []string
    Count int
}

opts := {
    Header: {"Name", "Email"}
    Funcs: {"{firstname} {lastname}", "{email}"} // Same as struct tags
    Count: 3
}

FixedWidth(opts)
```

```text
Name           Email
Billy White    billy@hotmail.com
Sally Fields   sally@gmail.com
Frank Bard     billy@hotail.com
```

I think this gets your FixedWidth function back in with the most versatility possible.

Let me know your thoughts. If your burnt out just let me know I can always wrap things up if you want.

Mrpye commented:
Sorry for the slowness in my response, I had an op on my hand so was out of action for a bit.

While waiting for pre-op, I was also thinking about the Implementation of the options (Waiting for an op and thinking about code :-}), It just felt a bit messy and confusing for the user with the lines in the TemplateOption. I came to the same conclusion having a data field and making the FuncMap public.
I then saw your message, separate Options, simple, I like it :-)
having the template parameter makes sense as it is a requirement, not an option.

Regarding the  FixedWidth function,  It's funny because generating FixedWidth data was the reason I started down the route with Templates, so happy to put it back in.

Another thing we need to consider is how to pass the width, alignment, and padding char of each field.
I have had use cases with fields padded out and aligned so it is needed.
Adding Spacing, Align, and Padding options to control this would work what are your thoughts?

```
opts := {
    Header: {"Name", "Email", "Account"}
    Spacing: {30,20,10}
    Align: {"",">","^"}
    Padding: {"","", "0"}
    Funcs: {"{firstname} {lastname}", "{email}", "{AchRouting}"} // Same as struct tags
    Count: 3
}

Name                        Email                                      Account
Billy White                 billy@hotmail.com      000123456789
Sally Fields                 sally@gmail.com         000123456789
Frank Bard                 billy@hotail.com          000123456789
```

I'll get started with the first lot of changes and wait for your thoughts on the extra options.

brianvoe commented:
We are on the same page! Agree with the first two parts and the third, just wanted to add my two cents to the fixed width part.

Totally agree with adding spacing, align and padding. I dont understand the align part but Im sure that makes sense to you. Is there a reason to have an array for those values instead of just a single value? and i just want to make sure those are optional and we have a default for those if they dont pass any in.

P.S. If you code this thing with one hand you are a badass!

Super duper close!

Mrpye commented:
Just a little update

Made loads of progress yesterday and got the fixed width working.
Just doing a bit of refinement and hardening.
Also got all the other changes in just need to update the docs.

I'll try to get some done today, but life getting in the way :-)



On Mon, 13 Nov 2023, 17:08 Brian Voelker, ***@***.***> wrote:

> We are on the same page! Agree with the first two parts and the third,
> just wanted to add my two cents to the fixed width part.
>
> Totally agree with adding spacing, align and padding. I dont understand
> the align part but Im sure that makes sense to you. Is there a reason to
> have an array for those values instead of just a single value? and i just
> want to make sure those are optional and we have a default for those if
> they dont pass any in.
>
> P.S. If you code this thing with one hand you are a badass!
>
> Super duper close!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/brianvoe/gofakeit/pull/266#issuecomment-1808604425>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABDIAXPATDO2KITZVJKJDNTYEJHZ3AVCNFSM6AAAAAA6Z75PFSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQMBYGYYDINBSGU>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

brianvoe commented:
No worries take your time.

Mrpye commented:
Hi,

So a summary of the latest push.
Template EmailText and Markdown now have their Options 
The template option has a Data field and the developer can access the function map.

The FixedWidth has been added.
I ended up changing the FixedWidth options from slices to fields, as it was getting complete validation
of the inputs. using fields has simplified the code and gives some consistency as it is very similar usage to the CSV function,

You can hide the header, footer
change the alignment of each column
change the padding char used for header, row, footer
set the column width or -1 auto calcs column width (default)
the function param  uses template functions or faker functions
also if a column has numeric data you can display the total in footer.

```go
// Output:
// name************last_nameEmail********************password****Account No.**Money
// Markus Moen     Daniel   marcelpagac@wuckert.biz  W8DAkpLjYoBW364599489953    35
// Amie Feil       Kuvalis  trystangislason@pagac.net9G9rOTgd3vDs635300425914    16
// Trystan GislasonKeebler  gabriellehuels@borer.io  J41S7H76KwYZ323202027613    61
//                                                                           112.00 
```

brianvoe commented:
Looking good! Make sure to think super simple, a lot of the times if something has too many options people are less prone to use it. Less is more type mentality. The separator per column and the footer are two examples of that.

Let me know when your done and ill review it then.

Thanks

Mrpye commented:
Ok, I have stripped it down to just spacing and alignment.
It defaults to left align and the spacing will auto-calc default.

brianvoe commented:
So much cleaner!!! Nice job. Sorry i dont mean to poo poo on your ideas I am just trying to give it the best possible success it can be. Cause i really want people to use this.

Mrpye commented:
I do get carried away :-),  You have a vision and it is quite a task making sure random people keep to it, I respect that. 
It's my first contribution to a project, and it's been a good learning experience.

brianvoe commented:
sounds good. let me know when your ready so we can get this in!!!

Mrpye commented:
I am ready

brianvoe commented:
ok im gonna merge this into develop and play around with it. Maybe find some way to simplify stuff.

brianvoe commented:
Ok first round I did cleanup and simplification of your fixed width function. Let me know what you think

https://github.com/brianvoe/gofakeit/commit/e3cd44073f814c1591387ab8857bcf624569f621

brianvoe commented:
@Mrpye  https://github.com/brianvoe/gofakeit/commit/082638fa6905e6a97d1c7eb115e4358ffd2e79dc

Ok i did a bunch of cleanup and get things in line with the rest of the package. Take a look at the commit.

What i need from you is to basically just get the test working again and get a decent amount of additional tests in so we know we are covering our bases. 

Take note that i have to use the rand that is passed to them. The rand they pass in need to be the bases for for how things are generated.

I tried building the FuncMap based upon the lookup setup but I think what you originally had by using reflect to pull the methods might be the better option.

I ran out of time today so ill let you run the finish line.

Just make sure you do a pr into the develop branch.

Mrpye commented:
Hi
I am working on the updates, I am nearly there, just have not had much time
this week.

Andrew

On Tue, Nov 21, 2023 at 10:56 PM Brian Voelker ***@***.***>
wrote:

> @Mrpye <https://github.com/Mrpye> 082638f
> <https://github.com/brianvoe/gofakeit/commit/082638fa6905e6a97d1c7eb115e4358ffd2e79dc>
>
> Ok i did a bunch of cleanup and get things in line with the rest of the
> package. Take a look at the commit.
>
> What i need from you is to basically just get the test working again and
> get a decent amount of additional tests in so we know we are covering our
> bases.
>
> Take note that i have to use the rand that is passed to them. The rand
> they pass in need to be the bases for for how things are generated.
>
> I tried building the FuncMap based upon the lookup setup but I think what
> you originally had by using reflect to pull the methods might be the better
> option.
>
> I ran out of time today so ill let you run the finish line.
>
> Just make sure you do a pr into the develop branch.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/brianvoe/gofakeit/pull/266#issuecomment-1821828537>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABDIAXPP6F7VI3SZ3WXGTQTYFUWSTAVCNFSM6AAAAAA6Z75PFSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQMRRHAZDQNJTG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

brianvoe commented:
no worries just let me know when your done

Mrpye commented:
Wicked very exciting 😀. It been learning experience. And thanks for
letting me contribute.


On Tue, 28 Nov 2023, 18:10 Brian Voelker, ***@***.***> wrote:

> @Mrpye <https://github.com/Mrpye> I submitted it to golang weekly
> newsletter and you made it first link!!!
>
> https://golangweekly.com/issues/486
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/brianvoe/gofakeit/pull/266#issuecomment-1830415045>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABDIAXOJSU7E2FBFEWN3Y2TYGYSLFAVCNFSM6AAAAAA6Z75PFSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQMZQGQYTKMBUGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
